DataDog/datadog-agent
# add new generic event platform aggregator API

### What does this PR do?

Add a new aggregator API through which checks can submit "event platform events" of various types.

All supported `eventTypes` are hardcoded in `EventPlatformForwarder`. 

The `dbm-samples` and `dbm-metrics` events are expected to arrive fully serialized so their pipelines are simply "HTTP passthrough" pipelines which skip all of the other features of logs pipelines like processing rules and encoding. 

Future event types will be able to add more detailed processing if they need it. 

#### Overall flow

1. `aggregator.submit_event_platform_event(check_id, rawEvent, "{eventType}")` - python API. Here's the corresponding integrations change to use the API: https://github.com/DataDog/integrations-core/pull/9165 
2. `BufferedAggregator` forwards events to the `EventPlatformForwarder`. Events are **dropped** here if `EventPlatformForwarder` is backed up for any reason. 
3. `EventPlatformForwarder` forwards events to the pipeline for the given `eventType`, **dropping** events for unknown `eventTypes`


#### Internal Agent Stats

*Prometheus*: `aggregator.flush - data_type:{eventType}, state:{ok|error}`

*ExpVar*: `EventPlatformEvents` & `EventPlatformEventsErrors`: counts by `eventType` 

#### User-Facing Agent Stats

Statistics for each `eventType` will be tracked alongside other types of telemetry (`Service Checks`, `Series`, ...). Where appropriate, the raw `eventType` is translated to a human readable name (i.e. `dbm-samples` --> `Database Monitoring Query Samples`). 

`agent status` output:
```
=========
Collector
=========

  Running Checks
  ==============

    postgres (5.4.0)
    ----------------
      Instance ID: postgres:1df52d84fb6f603c [OK]
      Metric Samples: Last Run: 366, Total: 7,527
      Database Monitoring Query Samples: Last Run: 11, Total: 176
      ...

=========
Aggregator
=========
  Checks Metric Sample: 29,818
  Database Monitoring Query Samples: 473
  ...
```

`agent check {check_name}` output:

```
=== Metrics ===
...
=== Database Monitoring Query Samples ===
...
```

`agent check {check_name} --json` output will use the raw event types instead of the human readable names:

```
"aggregator": {
  "metrics": [...],
  "dbm-samples": [...],
  ...
}
```

### Motivation

The posting of statement samples payloads to the intake for postgres & mysql checks is done directly from python as of (https://github.com/DataDog/integrations-core/pull/8627, https://github.com/DataDog/integrations-core/pull/8629). With this change we'll be able to move responsibility for posting payloads to the more robust agent go code with proper batching, buffering, retries, error handling, and tracking of statistics. 

### Describe your test plan

1. Run an agent build with https://github.com/DataDog/integrations-core/pull/9045 included
    1. Configure the postgres or mysql check to collect statement samples as described in https://github.com/DataDog/integrations-core/pull/8627 
    1. Enable agent telemetry `telemetry.enabled: true`
1. Confirm that statement samples are being ingested and appear on the database samples explorer
1. Manually run a mysql or postgres check (i.e. `agent check postgres`) and confirm that `Database Monitoring Query Samples` appear in the output, properly json formatted.
1. Run `agent status` and confirm that `Database Monitoring Query Samples` appear in the `postgres` check output as well as the `Aggregator` summary output. 
1. Run `agent launch-gui` and confirm that `Database Monitoring Query Samples` appear in both the `Aggregator` summary and the postgres or mysql check summary. 
1. Validate telemetry: `curl http://localhost:5000/telemetry | grep dbm-samples` to confirm that `dbm-samples` are appearing in the aggregator for `aggregator__flush` and `aggregator__processed` metrics
1. Validate expvar: Query `curl http://localhost:5000/debug/vars | grep dbm-samples` and confirm that `dbm-samples` are tracked correctly.
